# ***ML-Assignment-4-Classification Problem***  

#      ***Breast-Cancer-Prediction***


***Import Libraries***

In [373]:
import numpy as np#is a library for numerical computations
import pandas as pd #A powerful library for data manipulation and analysis
from sklearn.datasets import load_breast_cancer#his imports the breast cancer dataset from sklearn.datasets
from sklearn.preprocessing import StandardScaler#This ensures all features are on a similar scale
from sklearn.model_selection import train_test_split#Splits the dataset into training and testing subsets
from sklearn.linear_model import LogisticRegression#a linear classifier for binary outcomes.
from sklearn.tree import DecisionTreeClassifier# Which splits data based on feature thresholds to classify observations.
from sklearn.ensemble import RandomForestClassifier#decision trees that improves accuracy and reduces overfitting.
from sklearn.svm import SVC#Imports Support Vector Machine (SVM) with kernel functionality
from sklearn.neighbors import KNeighborsClassifier#a simple and non-parametric classifier.
#accuracy_score --Measures the overall accuracy of the model (correct predictions/total predictions).
#classification_report--Provides a detailed summary of precision, recall, and F1-score for each class.
#confusion_matrix---Evaluates performance by comparing actual vs. predicted labels in a matrix format
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix


# ***1.	Loading and Preprocessing*** 

***Loading and Preprocessing--
Load the breast cancer dataset from sklearn using the load_breast_cancer function from sklearn***

In [376]:
from sklearn.datasets import load_breast_cancer
cancer = load_breast_cancer()

In [378]:
df = pd.DataFrame(cancer.data, columns=cancer.feature_names)
df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

In [380]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

In [382]:
# information about the dimensions of the DataFrame, specifically the number of rows and columns.
df.shape

(569, 30)

In [384]:
#to print the number of missing (NaN) values in each column of the DataFrame
print(df.isnull().sum())

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
dtype: int64


***There is no missing values in the dataset,so need not imputation for this dataset.***

In [387]:
#to print the number of missing (NaN) values in each column of the DataFrame
df.dtypes

mean radius                float64
mean texture               float64
mean perimeter             float64
mean area                  float64
mean smoothness            float64
mean compactness           float64
mean concavity             float64
mean concave points        float64
mean symmetry              float64
mean fractal dimension     float64
radius error               float64
texture error              float64
perimeter error            float64
area error                 float64
smoothness error           float64
compactness error          float64
concavity error            float64
concave points error       float64
symmetry error             float64
fractal dimension error    float64
worst radius               float64
worst texture              float64
worst perimeter            float64
worst area                 float64
worst smoothness           float64
worst compactness          float64
worst concavity            float64
worst concave points       float64
worst symmetry      

***check the number of duplicate rows***

In [390]:
# to check the number of duplicate rows in the Dataframe
duplicates_count = df.duplicated().sum()

# Print the count of duplicate records
if duplicates_count > 0:
    print(f"\nNumber of duplicate records: {duplicates_count}\n")
else:
    print("Number of  duplicate records found.")
    
# Check for duplicates
duplicates = df[df.duplicated()]
# Print duplicates with header and all data
if not duplicates.empty:
    print(duplicates.to_string(index=False))
else:
    print("No duplicate records found.")


Number of  duplicate records found.
No duplicate records found.


In [392]:
df.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst radius  \
count     569.000000              569.000000  ...    569.000000   
mean        0.181162                0.062798  ...     16.269190   
std         0.027414                0.007060  ...      4.833242   
min         0.106000                0.049960  ...      7.930000   
25%         0.161900                0.057700  ...     13.010000   
50%         0.179200                0.061540  ...     14.970000   
75%         0.195700                0.066120  ...     18.790000   
max         0.304000                0.097440  ...     36.040000   

       worst texture  worst perimeter   worst area  worst smoothness  \
count     569.000000       569.000000   569.000000        569.000000   
mean       25.677223       107.261213   880.583128          0.132369   
std         6.146258        33.602542   569.356993          0.022832   
min        12.020000        50.410000   185.200000          0.071170   
25%        21.080000        84.110000   515.300000          0.116600   
50%        25.410000        97.660000   686.500000          0.131300   
75%        29.720000       125.400000  1084.000000          0.146000   
max        49.540000       251.200000  4254.000000          0.222600   

       worst compactness  worst concavity  worst concave points  \
count         569.000000       569.000000            569.000000   
mean            0.254265         0.272188              0.114606   
std             0.157336         0.208624              0.065732   
min             0.027290         0.000000              0.000000   
25%             0.147200         0.114500              0.064930   
50%             0.211900         0.226700              0.099930   
75%             0.339100         0.382900              0.161400   
max             1.058000         1.252000              0.291000   

       worst symmetry  worst fractal dimension  
count      569.000000               569.000000  
mean         0.290076                 0.083946  
std          0.061867                 0.018061  
min          0.156500                 0.055040  
25%          0.250400                 0.071460  
50%          0.282200                 0.080040  
75%          0.317900                 0.092080  
max          0.663800                 0.207500  

[8 rows x 30 columns]

# Breast cancer wisconsin (diagnostic) dataset


**Data Set Characteristics:**

***:Number of Instances: 569***

***:Number of Attributes: 30 numeric, predictive attributes and the class***

***:Attribute Information:
    - radius (mean of distances from center to points on the perimeter)  
    - texture (standard deviation of gray-scale values)  
    - perimeter  
    - area  
    - smoothness (local variation in radius lengths)  
    - compactness (perimeter^2 / area - 1.0)  
    - concavity (severity of concave portions of the contour)  
    - concave points (number of concave portions of the contour)  
    - symmetry  
    - fractal dimension ("coastline approximation" - 1)***  

***The mean, standard error, and "worst" or largest (mean of the three
    worst/largest values) of these features were computed for each image,
    resulting in 30 features.  For instance, field 0 is Mean Radius, field
    10 is Radius SE, field 20 is Worst Radius.***

***- class:
        - WDBC-Malignant  
            - WDBC-Benign***   



            
***The above details are available by executing the code block "print(cancer.DESCR)"***

In [395]:
cancer.target_names

array(['malignant', 'benign'], dtype='<U9')

In [397]:
print(cancer.feature_names)

['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']


# ***2. Classification Algorithm Implementation***

In [400]:
# Features and target for binary classification
X = cancer.data  # Features
y = cancer.target  # Target (binary class labels)


***Standarize the feature***

In [403]:
# Standardize the features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Split the data into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# Print the sizes of the train and test sets
print("Train set size:", X_train.shape)
print("Test set size:", X_test.shape)


Train set size: (455, 30)
Test set size: (114, 30)


# ***3. Logistic Regression***

In [406]:
# Logistic Regression
clf_logistic_regression = LogisticRegression(max_iter=10000, random_state=42)
clf_logistic_regression.fit(X_train, y_train)
y_pred_logistic_regression = clf_logistic_regression.predict(X_test)
accuracy_logistic_regression = accuracy_score(y_test, y_pred_logistic_regression)
print(f"Logistic Regression Accuracy: {accuracy_logistic_regression}")
print("Logistic Regression Classification Report:")
print(classification_report(y_test, y_pred_logistic_regression))
print("Logistic Regression Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_logistic_regression))

Logistic Regression Accuracy: 0.9736842105263158
Logistic Regression Classification Report:
              precision    recall  f1-score   support

           0       0.98      0.95      0.96        43
           1       0.97      0.99      0.98        71

    accuracy                           0.97       114
   macro avg       0.97      0.97      0.97       114
weighted avg       0.97      0.97      0.97       114

Logistic Regression Confusion Matrix:
[[41  2]
 [ 1 70]]


***The confusion matrix is a tool used to evaluate the performance of a classification model.***  
***Confusion Matrix Analysis***  

***considering  
                0 as malignant cases  
                1 as benign case***
    
    
***41    2   ----  The model correctly predicted 41 malignant cases, the model incorrectly  predicted 2 bengin case as malignant***  
***1     70  ----  The model incorrectly  predicted 1 benign case as malignant, the mode predicted 70 as benign case*** 

***Logistic Regression  
Logistic regression is a linear model that predicts the probability of a binary outcome (e.g., malignant or benign). It applies the sigmoid function to map predicted values to probabilities.  
The breast cancer dataset is well-structured, and logistic regression performs well when the relationship between features and the target is mostly linear.*** 

# ***4.Decision Tree Classifier***

In [411]:
# Decision Tree Classifier
clf_decision_tree = DecisionTreeClassifier(random_state=42)
clf_decision_tree.fit(X_train, y_train)
y_pred_decision_tree = clf_decision_tree.predict(X_test)
accuracy_decision_tree = accuracy_score(y_test, y_pred_decision_tree)
print(f"Decision Tree Classifier Accuracy: {accuracy_decision_tree}")
print("Decision Tree Classifier Classification Report:")
print(classification_report(y_test, y_pred_decision_tree))
print("Decision Tree Classifier Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_decision_tree))

Decision Tree Classifier Accuracy: 0.9473684210526315
Decision Tree Classifier Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.93      0.93        43
           1       0.96      0.96      0.96        71

    accuracy                           0.95       114
   macro avg       0.94      0.94      0.94       114
weighted avg       0.95      0.95      0.95       114

Decision Tree Classifier Confusion Matrix:
[[40  3]
 [ 3 68]]


***Confusion Matrix Analysis***        
***40    3   ----  The model correctly predicted 40 malignant cases, the model incorrectly  predicted 3 bengin case as malignant***  
***3     68  ----  The model incorrectly  predicted 3 benign case as malignant, the mode predicted 68 as benign case*** 

 ***Decision Tree Classifier***  
 ***Decision trees split the data into branches based on feature thresholds, creating a flowchart-like structure to classify observations.***

***Decision trees can handle both linear and non-linear relationships, making them versatile for this dataset. They are also interpretable, which is valuable in medical diagnostics.***

# ***5. Random Forest Classifier***

In [416]:
#Random Forest Classifier
clf_random_forest = RandomForestClassifier(n_estimators=100, random_state=42)
clf_random_forest.fit(X_train, y_train)
y_pred_random_forest = clf_random_forest.predict(X_test)
accuracy_random_forest = accuracy_score(y_test, y_pred_random_forest)
print(f"Random Forest Classifier Accuracy: {accuracy_random_forest}")
print("Random Forest Classifier Classification Report:")
print(classification_report(y_test, y_pred_random_forest))
print("Random Forest Classifier Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_random_forest))

Random Forest Classifier Accuracy: 0.9649122807017544
Random Forest Classifier Classification Report:
              precision    recall  f1-score   support

           0       0.98      0.93      0.95        43
           1       0.96      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.97      0.96      0.96       114
weighted avg       0.97      0.96      0.96       114

Random Forest Classifier Confusion Matrix:
[[40  3]
 [ 1 70]]


***Confusion Matrix Analysis***    
***40    3   ----  The model correctly predicted 40 malignant cases, the model incorrectly  predicted 3 bengin case as malignant***  
***1     70  ----  The model incorrectly  predicted 1 benign case as malignant, the mode predicted 70 as benign case*** 

 ***Random Forest is an ensemble of decision trees. It builds multiple trees on random subsets of the data and aggregates their predictions (majority voting for classification).*** 

***This method reduces overfitting, which can occur with a single decision tree. It’s robust and often achieves high accuracy on structured datasets like this one.*** 

# ***6. Support Vector Machine (SVM)***

In [421]:
# Support Vector Machine (SVM)
clf_svm = SVC(random_state=42)
clf_svm.fit(X_train, y_train)
y_pred_svm = clf_svm.predict(X_test)
accuracy_svm = accuracy_score(y_test, y_pred_svm)
print(f"SVM Accuracy: {accuracy_svm}")
print("SVM Classification Report:")
print(classification_report(y_test, y_pred_svm))
print("SVM Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_svm))

SVM Accuracy: 0.9736842105263158
SVM Classification Report:
              precision    recall  f1-score   support

           0       0.98      0.95      0.96        43
           1       0.97      0.99      0.98        71

    accuracy                           0.97       114
   macro avg       0.97      0.97      0.97       114
weighted avg       0.97      0.97      0.97       114

SVM Confusion Matrix:
[[41  2]
 [ 1 70]]


***Confusion Matrix Analysis***        
***41    2   ----  The model correctly predicted 41 malignant cases, the model incorrectly  predicted 2 bengin case as malignant***  
***1     70  ----  The model incorrectly  predicted 1 benign case as malignant, the mode predicted 70 as benign case*** 

 ***SVM finds a hyperplane that best separates the data into two classes. For non-linearly separable data, it uses kernels (e.g., radial basis function) to map data to higher dimensions.***  
***SVM works well on datasets with clear class boundaries and is effective for high-dimensional data, making it a strong choice for the 30-feature breast cancer dataset.***

#  ***7. k-Nearest Neighbors (k-NN)***

In [426]:
#k-Nearest Neighbors (k-NN)
clf_knn = KNeighborsClassifier(n_neighbors=5)
clf_knn.fit(X_train, y_train)
y_pred_knn = clf_knn.predict(X_test)
accuracy_knn = accuracy_score(y_test, y_pred_knn)
print(f"k-NN Accuracy: {accuracy_knn}")
print("k-NN Classification Report:")
print(classification_report(y_test, y_pred_knn))
print("k-NN Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_knn))

k-NN Accuracy: 0.9473684210526315
k-NN Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.93      0.93        43
           1       0.96      0.96      0.96        71

    accuracy                           0.95       114
   macro avg       0.94      0.94      0.94       114
weighted avg       0.95      0.95      0.95       114

k-NN Confusion Matrix:
[[40  3]
 [ 3 68]]


***Confusion Matrix Analysis***    
***40    3   ----  The model correctly predicted 40 malignant cases, the model incorrectly  predicted 3 bengin case as malignant***  
***3     68  ----  The model incorrectly  predicted 3 benign case as malignant, the mode predicted 68 as benign case*** 

***k-Nearest Neighbors (k-NN)***  
***k-NN classifies a data point based on the class of its k nearest neighbors in the feature space. It’s a non-parametric algorithm.***

***k-NN is straightforward to implement and can perform well when the dataset is not too large (as in this case). It adapts naturally to non-linear decision boundaries.***

# ***8. Compare the performance of the five classification algorithms. Which algorithm performed the best and which one performed the worst?***

In [431]:
# Find the best and worst models based on accuracy
best_model = max(results, key=lambda x: x['accuracy'])
worst_model = min(results, key=lambda x: x['accuracy'])

print(f"The best performing model is: {best_model['model']} with an accuracy of {best_model['accuracy']:.4f}")
print(f"The worst performing model is: {worst_model['model']} with an accuracy of {worst_model['accuracy']:.4f}")

The best performing model is: LogisticRegression with an accuracy of 0.9737
The worst performing model is: DecisionTreeClassifier with an accuracy of 0.9474
